### PR TITLE
Add WaveShaperNode

### PR DESCRIPTION
**Pull request by Jerboas86**
- Add `WaveShaperNode` ands its rendering part `WaveShaperRenderer`. All oversampling modes are supported
- Add simple tests
- Add example shapper.rs

For the oversampling, i used [rubato crate](https://docs.rs/rubato/0.10.0/rubato/index.html).

related to #25

### CODE REVIEW

File: src/node/waveshaper.rs [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+
+        for (i_data, o_data) in wave_out.iter().zip(output.channels_mut()) {
+            for (&i, o) in i_data.iter().zip(o_data.iter_mut()) {
+                *o = i;
+            }
+        }
+    }
+
+    #[inline]
+    fn tick(&self, input: f32) -> f32 {
```
[the remark this answers is not in the document]
Comment (orottier) [reply src/node/waveshaper.rs:439]: I assume 
- there are a lot branches (and they cannot reliably be predicted over time)
- random array indexing of self.curve (which can be large so not very cacheable)

The only optimization I can think of now is making self.curve smaller in memory, so it allows better caching. E.g. make is a `Vec<f16>`. That probably is not compatible with the spec though..

File: src/node/waveshaper.rs [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+        let n = self.curve.len() as f32;
+        if n == 0. {
+            return 0.;
+        }
+        let v = (n - 1.) / 2.0 * (input + 1.);
+
+        match v {
```
Comment (orottier) [inline src/node/waveshaper.rs:446]: This match statement is rather odd, you are matching `v` to itself. Better to write it as a set of if-statements.

File: src/node/waveshaper.rs [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+            for (&i, o) in i_data.iter().zip(o_data.iter_mut()) {
+                *o = i;
+            }
+        }
+    }
+
+    #[inline]
+    fn tick(&self, input: f32) -> f32 {
+        let n = self.curve.len() as f32;
+        if n == 0. {
```
Comment (orottier) [inline src/node/waveshaper.rs:441]: Better to move this check up one line, and rewrite is as `if self.curve.is_empty()`

File: src/node/waveshaper.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (orottier) [reply src/node/waveshaper.rs:439]: Also, in your example you use a 441 length shape. That is pretty big. Try a size of 100, that will allow for precision of signal of 0.01, right? Enough precision for a human ear

**Comment by Jerboas86** [comment]
I changed OverSampleType::X4 to OverSampleType::None to make the example run without underrun in debug mode.

File: src/node/waveshaper.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jerboas86) [reply src/node/waveshaper.rs:439]: > That probably is not compatible with the spec though

Indeed the spec request a [f32 array for curve](https://www.w3.org/TR/webaudio/#dom-waveshapernode-curve)

**Pull request merged**